File: DSU/DSU.cpp
class DSU {
private:
    vector<int> P, SZ;
    int components;

public:
    DSU(int n): components(n) {
        P.resize(n);
        SZ.assign(n, 1);
        iota(P.begin(), P.end(), 0);
    }

    bool union(int x, int y) {
        x = find(x), y = find(y);

        if (x == y) {
            return false;
        }

        if (SZ[x] > SZ[y]) {
            swap(x, y);
        }

        P[x] = y;
        SZ[y] += SZ[x];
        components--;
        return true;
    }

    int find(int x) {
        return P[x] == x ? x : P[x] = find(P[x]);
    }
};

File: DSU/DSU.py
class DSU:
    def __init__(self, N):
        self.N = N
        self.comps = N
        self.parent = list(range(N))
        self.size = [1] * N

    def find(self, X):
        if self.parent[X] != X:
            self.parent[X] = self.find(self.parent[X])
        return self.parent[X]

    def union(self, X, Y):
        X, Y = self.find(X), self.find(Y)
        if X == Y:
            return False
        if self.size[X] > self.size[Y]:
            X, Y = Y, X
        self.parent[X] = Y
        self.size[Y] += self.size[X]
        self.comps -= 1
        return True
